docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,295 @@
+As an IT administrator, I want the name of my MicroShift instance in ACM to have the same cluster name as the underlying device's name or fingerprint, so I can easily correlate between the two.
+
+#### Story 5: site specific telemetry endpoints
+As an IT administrator, I want to provide site-specific/region-specific telemetry endpoints to devices.
+
+#### Story 6: dialing up verbosity for logging in my user application in specific devices
+As an IT administrator, I want to dial up the verbosity for logging on a specific device, so I can troubleshoot it.
+
+#### Story 7: region specific application service endpoints
+As an IT administrator, I want to provide region-specific service endpoints to my applications
+
+#### Story 8: region specific radio bands/regulations
+As a developer, I want to ensure that my devices automatically comply with local regulations when enrolled in specific regions. (i.e. configured with region-specific radio bands for WiFi, 5G, or other radio technologies)
+
+#### Story 9: region specific data sovereignty
+As a developer, I want to ensure that my devices connect to the specific regional endpoints to ensure compliance with data sovereignty.
+
+### Notes/Constraints/Caveats
+
+We would need to identify the design changes relative to templateVersions and
+if the relationship of those continues to be to the fleet or device level.
+
+#### Reference freezing drift
+
+One of the issues identified with this design, is that it's harder to pre-freeze the refences that would potentially
+surface from rendering all devices. For example if we had:
+
+```yaml
+os:
+  image: forklifts-{{ .device.metadata.label[factory] }}:latest
+```
+
+we couldn't know ahead of time what the image would be, and we would need to render all devices to know what the
+image would be. External git references with variable templating would have the same issue.
+
+Not being able to pre-freeze references while performing a roll-out means that some devices would potentially
+have one image, and if during rollout the image tag get's updated, the remaining devices will have a
+different one, while the referencing labels are still the same.
+
+##### Potential solutions
+
+1. We ignore it, and expect eventual consistency, on a 2nd round the server will detect that the image
+   reference has been updated, and the devices with the older images will be re-rendered.
+   Additionally, when using free-floating tags, it's probably a risk the customer is assuming in
+   favor of flexilibity and speed of development. This should in general be a practice we don't
+   recommend, as we only supporting it to make development easier while unfavouring determinism.
+
+2. We could mantain a list of all possible values for labels possibilities referenced within a fleet
+   This would help us create a list of references (i.e. an internal database object that holds all references)
+   for a rollout. This would not help with device names, but would help with labels.
+
```

Comment:
3. After variable substitution, whenever we encounter a given (config_source, floating_tag)-tuple for the first time we cache a snapshot locally and use that same substitution for all following devices with the same tuple. This ensures consistency for all like devices. Devices with another tuple (after substituting in labels) will anyway receive a different configuration. Then we apply the same rules about "expiring" cache entries and re-checking sources as we have to do without variable substitution anyway.